### PR TITLE
merge bitcoin#15043, #15295, #15399, #17452, #16338, #15504, #17076, #17069, #17018: fuzzing

### PR DESCRIPTION
**Pull request by kittywhiskers**

### CODE REVIEW

**Review by PastaPastaPasta — COMMENTED** [review]
partial review, almost everything looks good

File: src/Makefile.test.include [hunk 1]
```diff
@@ -55,14 +55,34 @@ RAW_TEST_FILES = \
 GENERATED_TEST_FILES = $(JSON_TEST_FILES:.json=.json.h) $(RAW_TEST_FILES:.raw=.raw.h)
 
 BITCOIN_TEST_SUITE = \
-  test/test_dash_main.cpp \
+  test/main.cpp \
```
Comment (UdjinM6) [inline src/Makefile.test.include:77]: 15504: should drop `test_dash_main.cpp` file

File: src/test/main.cpp [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Copyright (c) 2011-2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#define BOOST_TEST_MODULE Bitcoin Core Test Suite
```
Comment (UdjinM6) [inline src/test/main.cpp:5]: 15504: nit: Dashify?

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

**Comment by PastaPastaPasta** [comment]
See https://github.com/PastaPastaPasta/dash/tree/pr_4312 for my rebased branch

File: src/Makefile.test.include [hunk 3]
```diff
@@ -148,28 +178,348 @@ test_test_dash_LDFLAGS = $(LDFLAGS_WRAP_EXCEPTIONS) $(RELDFLAGS) $(AM_LDFLAGS) $
...
-  $(LIBSECP256K1)
-
-test_test_dash_fuzzy_LDADD += $(BOOST_LIBS) $(CRYPTO_LIBS) $(BACKTRACE_LIB)
-#
+
+if ENABLE_FUZZ
+test_fuzz_block_deserialize_SOURCES = $(FUZZ_SUITE) test/test_dash_fuzzy.cpp
+test_fuzz_block_deserialize_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES) -DBLOCK_DESERIALIZE=1
+test_fuzz_block_deserialize_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
+test_fuzz_block_deserialize_LDFLAGS = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
```
Comment (UdjinM6) [inline src/Makefile.test.include:227]: missing `$(LDFLAGS_WRAP_EXCEPTIONS)`? (same for other `_LDFLAGS`s below)

File: src/Makefile.test.include [hunk 4]
```diff
@@ -148,28 +178,348 @@ test_test_dash_LDFLAGS = $(LDFLAGS_WRAP_EXCEPTIONS) $(RELDFLAGS) $(AM_LDFLAGS) $
...
+ $(LIBBITCOIN_SERVER) \
+ $(LIBBITCOIN_COMMON) \
+ $(LIBBITCOIN_UTIL) \
+ $(LIBBITCOIN_CONSENSUS) \
+ $(LIBBITCOIN_CRYPTO) \
+ $(LIBBITCOIN_CRYPTO_SSE41) \
+ $(LIBBITCOIN_CRYPTO_AVX2) \
+ $(LIBBITCOIN_CRYPTO_SHANI) \
+ $(LIBSECP256K1)
+test_fuzz_block_deserialize_LDADD += $(BOOST_LIBS) $(CRYPTO_LIBS)
```
Comment (UdjinM6) [inline src/Makefile.test.include]: missing `$(BACKTRACE_LIB)`? (same for other `_LDADD`s below)

File: .travis.yml [hunk 5]
```diff
@@ -281,6 +282,17 @@ after_success:
...
+      env: >-
+        HOST=x86_64-unknown-linux-gnu
+        PACKAGES="clang llvm python3 libssl1.0-dev libevent-dev bsdmainutils libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev"
+        NO_DEPENDS=1
+        RUN_UNIT_TESTS=false
+        RUN_FUNCTIONAL_TESTS=false
+        RUN_FUZZ_TESTS=true
+        GOAL="install"
+        BITCOIN_CONFIG="--disable-wallet --disable-bench --with-utils=no --with-daemon=no --with-libs=no --with-gui=no --enable-fuzz --with-sanitizers=fuzzer,address CC=clang CXX=clang++"
+
```
Comment (UdjinM6) [inline .travis.yml:295]: would be nice to have a Gitlab CI job similar to this to make sure things work as expected

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
- 15295 is partial I think
- linter complains "Good job!" :) https://gitlab.com/dashpay/dash/-/jobs/1486652462#L57
- see below

File: src/Makefile.test.include (hunk 3 above)
Comment (kittywhiskers) [reply src/Makefile.test.include]: Resolved

File: src/Makefile.test.include (hunk 4 above)
Comment (kittywhiskers) [reply src/Makefile.test.include]: Resolved

**Comment by UdjinM6** [comment]
linter still complains
```
Good job! The Boost dependency "boost/test/unit_test_monitor.hpp" is no longer used.
Please remove it from EXPECTED_BOOST_INCLUDES in test/lint/lint-includes.sh
```
https://gitlab.com/dashpay/dash/-/jobs/1493947535#L57

**Review by UdjinM6 — APPROVED** [review]
LGTM, utACK 👍

File: src/test/test_dash_fuzzy.cpp [hunk 6]
```diff
@@ -1,11 +1,7 @@
-// Copyright (c) 2009-2015 The Bitcoin Core developers
+// Copyright (c) 2009-2018 The Bitcoin Core developers
```
Comment (PastaPastaPasta) [inline src/test/test_dash_fuzzy.cpp]: nit: this shouldn't have been changed

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**